### ISSUE TITLE
[WIP] [RFC] Remove hard-coded vector column requirements

### ISSUE THREAD

**Issue by eddyxu** [lancedb/lancedb#462]
# Problem Statement
`LanceDB` currently imposes a hard limitation about what the schema needs to have for a vector column.

It became the root cause of multiple issues:

#96 
#360
#68 
#459 
#357

# Proposed changes

TBD

**Comment by changhiskhan** [comment]
1. What would the search API look like for FTS?
2. What would the search API look like with a serialized embedding function?

**Comment by changhiskhan** [comment]
Something like this?

```
Vector = Union[list, tuple, np.ndarray, pa.array]

class Table:
   def search(query: str | Vector,  column: str | None = None, query_type: str = "vector"):
        ....
```

Vector search with user created embedding:
```
table.search([1.2, 5.4])
```

Vector search with user created embedding with multiple vector columns of same length:
```
table.search([1.2, 5.4], column="foo")
```

Vector search with a serialized embedding function:
```
table.search("search query text")
```

Vector search with a serialized embedding function with more than 1 vector column:
```
table.search("search query text", column="foo")
```

FTS with a single index
```
table.search("search query text", query_type="fts")
```

Future: hybrid
```
table.search("search query text", query_type="hybrid", weight=0.5)
```

**Comment by eddyxu** [comment]
The search API between text w/embeddings and full text search are too similar tho.

How about this

```python
table.query().vector([1, 2, 3, 4], /* column="foo" */)
table.query().vector("input text", /* column="foo" */)
```

```python
table.query().fts("full text search", column="foo")
```

In future, the hybrid search can be

```
table.query()
  .vector([1, 2, 3], column="embed", weight=0.2)
  .fts("text search", column="text", weight=0.73)
  
```

**Comment by westonpace** [comment]
> table.query().vector("input text", /* column="foo" */)

Is this the user created embedding case where "input text" is converted to an embedding?

I suppose, with user-defined embeddings, the first argument could be any python object (e.g. image, sound file, etc.)

**Comment by eddyxu** [comment]
>Is this the user created embedding case where "input text" is converted to an embedding?

Yes, this should be able to take anything that embedding function takes. It does raise a question how to handle such parameter in rust / javascript tho.

**Comment by changhiskhan** [comment]
nit: we're flipping back and forth between `query` and `search`. Let's just pick one and stick with it (or maybe use an alias).

> ```python
> table.query().vector([1, 2, 3, 4], /* column="foo" */)
> table.query().vector("input text", /* column="foo" */)
> ```
> 
> ```python
> table.query().fts("full text search", column="foo")
> ```

So a single FTS index can span multiple columns. So the "column" parameter here is somewhat confusing.


> 
> In future, the hybrid search can be
> 
> ```
> table.query()
>   .vector([1, 2, 3], column="embed", weight=0.2)
>   .fts("text search", column="text", weight=0.73)
>   
> ```

This is a fairly major refactor with heavy breaking API changes. What would the migration process look like?

**Comment by eddyxu** [comment]
> So a single FTS index can span multiple columns. So the "column" parameter here is somewhat confusing.

Maybe `columns` to `fts()`, it makes more sense to separate vector / fts to different functions in this regards, because other wise there will be `column=None, columns=None` in the same signature. 

>This is a fairly major refactor with heavy breaking API changes. What would the migration process look like?

This is backward compatible to the current API, as `search()` is what we have now, and `query()...` is a query build pattern for more complicated queries.  

Implementation wise, we can just make `search(..)` to call `query().....execute()`.  And mark the difference as "simple API vs more controlled APIs",  the query builder API is more familiar in sqlalchemy / db APIs. 

Long term, we can decide whether we wanted to put a deprecate warning in `search()` API from user feedbacks.

**Comment by AyushExel** [comment]
Just catching up with this.
So, this is the current plan? 
For higher level - `table.search(list-like)` for vector search & `table.search(str)` for fts
For advanced query building - `table.query().vector(list-like or "str", ...)`  & `table.query().fts(str, ...)`

A couple thoughts come to mind:
- How are we treating str inputs for vector search in high level search() function. It's not clear rn as the syntax clashes with high level search api for fts, no?
- "table.query()" is common in both advanced usage APIs, so can we just eliminate it and do something this:
    `tbl.vector(...)` -> internally calls self.query()... query builder `# advanced query builder for vector search`
    `tbl.fts(...)` -> internally calls self.query().. query builder `# advanced query builder for fts`
 Although tbl.vector seems odd as it's not a verb like fts() or search()
